### ISSUE TITLE
Implement tessellation functionality in Godot

### ISSUE THREAD

**Issue by Arnklit** [godotengine/godot-proposals#5995]
### Describe the project you are working on

Various 3D projects, including the Waterways add-on

### Describe the problem or limitation you are having in your project

It's difficult or impossible to manage high density details in Godot at the moment. Especially with non-static meshes such as water, hair, deforming snow.

While POM can help with adding fine detail it has a lot of limitations:
* POM doesn't change profile of the mesh, causing the effect to fall apart at certain angles.
* POM doesn't cast shadows from the added displacement (this might be solved).
* POM displacement doesn't intersect correctly, causing the effect to fall apart when objects are places on top of or intersecting with the POM surface.
* POM doesn't work with tri-planar texture mapping, limiting it's use for environments (this might be solved).

While Godot 4 can handle a lot more geometry than Godot 3.x, I don't think it's feasible to add all this detail into the meshes for most users. Even if Godot eventually gets some kind of virtualised geometry solution similar to "Nanite", it is often much easier to add details with tiling textures than using dense meshes, both from an asset creation point of view and dealing with the asset import / storage.

### Describe the feature / enhancement and how it helps to overcome the problem or limitation

An option to enable dynamic tessellation on a material or mesh.

### Describe how your proposal will work, with code, pseudo-code, mock-ups, and/or diagrams

I am aware that geometry shaders are getting deprecated, so this would have to be a custom compute shader implementation, but I do not know the details on how this would be done.

### If this enhancement will not be used often, can it be worked around with a few lines of script?

This would be used often and cannot be done easily with a few lines of script.

### Is there a reason why this should be core and not an add-on in the asset library?

This has to be done in core.

**Comment by Arnklit** [comment]
Note I am by no means a graphics programmer, but researching this a bit more, it seems it is possible to implement into the core of Godot 4.0's Vulkan renderer without geometry shaders by implementing tessellation stages.

https://registry.khronos.org/vulkan/specs/1.3-khr-extensions/html/chap22.html

**Comment by Calinou** [comment]
Tesselation has lost a lot of traction over the years, in favor of using actual geometry to represent bumps in surfaces where it matters. A lot of GPUs implement tesselation in a slow manner, making it "not worth it" to use versus actual geometry with a LOD system.

**Comment by Arnklit** [comment]
@Calinou do you have any details on the performance thing with NVidia? Wouldn't that be when using geometry shaders?

There are modern examples of tessellation made for consoles only, which are AMD chips. Like the Demon's Souls remaster.
https://youtu.be/kAv-ajC1bWg?t=365

**Comment by myaaaaaaaaa** [comment]
Just thought I'd chime in with a more fleshed out design draft, since I've been contemplating the best way to implement tessellation myself and it seems like there's a reasonable amount of interest in it.

Although, right now I'm leaning more towards LODs for my particular use case, but I thought I'd at least share what I came up with so far, if only as a springboard for someone else to jump off of.

### Basic usage
To quickly get started with tessellation, the developer simply needs to add a single line to their GDShader while making sure the attached mesh is formatted as PrimitiveType.PRIMITIVE_TRIANGLES:
```
shader_type spatial;
render_mode patch3;

void vertex() {
    ...
}
void fragment() {
    ...
}
```
32 new render_modes (patch1-patch32) will be made available, one for every supported patch size (although I'm not too sure about the necessity of this - OpenGL/Vulkan/DirectX allow patch sizes up to 32, but I feel like only triangles and quads are relevant in practice?)

On supported hardware, setting a patch size with render_mode will enable tessellation for that GDShader and cause it to override the primitive of the mesh it's attached to, rendering it as a VK_PRIMITIVE_TOPOLOGY_PATCH_LIST/GL_PATCH using the specified patch size. Otherwise, the tessellation functionality will be ignored and the GDShader will revert to being a normal vertex+fragment shader.

Theoretically, since PrimitiveType.PRIMITIVE_TRIANGLES shares the same vertex format as patch3, using that combination should allow developers to enable tessellation in a way that easily falls back to a normal mesh on unsupported hardware. However, this probably needs investigation since I'm not sure how GL_PATCH interacts with index buffers.

Warning messages should probably be emitted if there is a patch size/PrimitiveType mismatch when running on unsupported hardware.

It's probably a good idea to make sure that all PrimitiveMeshes use PrimitiveType.PRIMITIVE_TRIANGLES to make it as easy as possible for developers to get started with tessellation.

### tessellate()
A new shader stage that runs before `vertex()`. If tessellation is enabled with `render_mode patchN`, it will run for every edge (both outer and inner) and write to a new built-in variable, TESSELLATION_FACTOR, which controls how much to tessellate that particular edge.

By default (if `tessellate()` is omitted), TESSELLATION_FACTOR is set to the screen-space length of the edge in pixels, which will cause the edge to be tessellated to roughly a pixel size of one. Specifying a `tessellate()` function will allow the shader author to override this.

Under-the-hood, this would probably require the tessellation control shader to transform the patch vertices using `vertex()` in order to derive the default TESSELLATION_FACTOR, but then pass untransformed vertices to the tessellation evaluation shader to call `vertex()` again.

This code will tessellate all edges to a pixel size of 2 to reduce triangle count 4x compared to a pixel size of 1:
```
shader_type spatial;
render_mode patch4;

void tessellate() {
    TESSELLATION_FACTOR /= 2.0;
}
```
This code will target a pixel size of one, but prioritize rounding the tessellation to an integer power of two. This makes it work similar to blender's subdivision levels, where new points are added without moving old points, which minimizes tessellation popping artifacts:
```
shader_type spatial;
render_mode patch4;

void tessellate() {
    TESSELLATION_FACTOR = exp2(round(log2(TESSELLATION_FACTOR)));
}
```
This code will ignore pixel sizes and tessellate all edges by a fixed amount (four in this case):
```
shader_type spatial;
render_mode patch4;

void tessellate() {
    TESSELLATION_FACTOR = 4.0;
}
```
This pseudocode will avoid tessellating backfacing patches. As this is run per-edge rather than per-vertex, WORLD_NORMAL should be set to the average of the two vertices for outer edges, and the average of all vertices in the patch for inner edges.

However, if this can be made to work seamlessly for everyone, it's better to do this automatically in the GLSL side. Patches outside the camera bounds are also candidates for being excluded from tessellation.
```
shader_type spatial;
render_mode patch4;

void tessellate() {
    if (dot(WORLD_NORMAL, CAMERA_DIRECTION_WORLD) > 0.1)
        TESSELLATION_FACTOR = 1.0;
}
```

**Comment by ja2142** [comment]
Tessellation being slow on non-nvidia GPUs is a valid concern, but I think there are cases in which this is the most sane approach: one of them is having a large geometry with fine details in which both close-up and far away elements may be visible at the same time (prime examples of this would be terrain, large bodies of water etc.). In those cases other methods I know either don't work or are rather hacky:
 - LOD works per mesh, that's not helpful,
 - dividing large mesh into chunks to take advantage of LOD wouldn't be that bad, except it generates more problems on chunk boundaries.
 - POM is kind of ok for adding really small details, but doesn't do anything to geometry, so say a low poly mountain occluding something farther away would look terrible.
 - There's a moving clipmap idea (https://www.youtube.com/watch?v=rcsIMlet7Fw) which, while very clever, requires moving a terrain, even if it is stationary, which spills rendering complexity outside of the shader itself.
 - Virtual geometry/nanite type of solution seems really nice, and it would solve most of the terrain problem. Main issue with this approach is that it isn't implemented in godot at the moment. Even if it was though, I'm not entirely sure how good it can work with animated or (partially) procedurally generated models.

So unless tessellation performance makes it unusable, I think it'd be better then all of the solutions above in cases like terrain and water.

@myaaaaaaaaa you propose to run tessellate() before vertex(). In both vulkan and opengl tessallation seems to be a step after vertex shader. So the pipeline probably has to be done more like in those apis:
```
vertex -> tessellation control -> tessallation evaluation -> fragment
```
Tessellation evaluation coule be named something like `interpolate()`?

There's a quite nice sample of tessellation shaders in https://github.com/KhronosGroup/Vulkan-Samples in case someone tries to implement this:
https://github.com/KhronosGroup/Vulkan-Samples/blob/master/samples/api/terrain_tessellation/terrain_tessellation.cpp
https://github.com/KhronosGroup/Vulkan-Samples/blob/master/shaders/terrain_tessellation/terrain.tesc
https://github.com/KhronosGroup/Vulkan-Samples/blob/master/shaders/terrain_tessellation/terrain.tese

**Comment by PeterSHollander** [comment]
I would like to add support for seeing tessellation implemented in Godot 🌐

In my use case, it would be a major fidelity boost for curved model silhouettes in XR content (see [PN-Triangles](https://alex.vlachos.com/graphics/CurvedPNTriangles.pdf) for reference), where it is not predictable how close the camera will ever be to any object, and where tessellation would save an otherwise unreasonable amount of LOD file size to achieve a similar result.  Additionally, as suggested [here](https://testing.godotforums.org/discussion/22945/looking-for-someone-to-implement-real-time-tessellation-of-subdivision-surfaces-for-a-stretch-goal), using realtime tessellation for a large amount of models would save significantly on memory usage.  A use case of tessellation shaders for skewed viewing angles of large meshes such as terrain can be found in #6279 ([comment](https://github.com/godotengine/godot-proposals/issues/6279#issuecomment-1427199950)) and #445 ([comment](https://github.com/godotengine/godot-proposals/issues/445#issuecomment-1374027384)).  Silhouette tessellation may also relate to the implementation of OpenSubdiv proposed in #784, but as it operates realtime on the CPU it is my understanding that GPU realtime tessellation would be more performant (and potentially more flexible for non-uniform tessellation).

I would also like to ask for clarification if there is a possibility to achieve tessellation-like behaviour using compute shaders, as suggested by @fire in #2075 ([comment](https://github.com/godotengine/godot-proposals/issues/2075#issuecomment-754699984)), and adjacently suggested in [this comment](https://github.com/godotengine/godot/pull/28237#issuecomment-485844909)?

It appears that @painfulexistence is [experimenting with adding tessellation](https://github.com/painfulexistence/godot/tree/tessellation), as mentioned in #6121 ([comment](https://github.com/godotengine/godot-proposals/issues/6121#issuecomment-1436005901)).  While I am not graphics engine programmer, I am willing to learn if it can help contribute the feature of tessellation to core Godot.

**Comment by myaaaaaaaaa** [comment]
> I would also like to ask for clarification if there is a possibility to achieve tessellation-like behaviour using compute shaders, as suggested by @fire in #2075 ([comment](https://github.com/godotengine/godot-proposals/issues/2075#issuecomment-754699984)), and adjacently suggested in [this comment](https://github.com/godotengine/godot/pull/28237#issuecomment-485844909)?
> 

[Mesh shaders](https://devblogs.microsoft.com/directx/coming-to-directx-12-mesh-shaders-and-amplification-shaders-reinventing-the-geometry-pipeline/) are the "official" modern way of creating geometry from compute shaders, since unlike compute shaders, mesh shaders don't need to make a round trip to VRAM to store geometry in an intermediate buffer.

Their downside compared to tessellation and compute shaders is that they're a relatively recent feature, supported only on NVIDIA RTX, Intel Arc, and Radeon RX6000.

**Comment by fire** [comment]
https://github.com/v-sekai/godot-subdiv needs a implementation of vulkan compute shaders.

See also https://github.com/PixarAnimationStudios/OpenSubdiv/issues/1153

**Comment by Igameart** [comment]
Just leaving my support for tessellation. It would be mostly beneficial, in my opinion, on terrain systems similar to how it's used in Unreal engines. I would love to see that implementation in Godot.

**Comment by erickweil** [comment]
> I am aware that geometry shaders are getting deprecated, so this would have to be a custom compute shader implementation, but I do not know the details on how this would be done.

Yes, It **could** be a compute shader implementation but the current way they work requires a sync read back to CPU which invalidates their usage for anything happening each frame.

_My scenario_
I'm trying to generate triangles in the fly and render then each frame (It's for showing the 3d slice of 4d geometry made of tetrahedrons, but the exact use-case is not relevant). At least in theory should be possible using compute shaders to produce geometry by storing the result in buffers (generated triangle Indices and Vertex Data) and then rendering it without creating a Mesh. 

I managed to get the compute shader part working, but no hopes in making the result go to the normal render pipeline like a normal Mesh would go, without actually reading the data back to CPU and creating the ArrayMesh. This basically turns compute shaders not useful for this kinds of tasks since the CPU syncs eats up any performance gains that we would otherwise gain.

In conclusion...  If compute shaders were allowed to 

- Write to Textures used by objects in the scene (Upcoming in 4.2 https://github.com/godotengine/godot-demo-projects/pull/938)
- Bind the same output buffer used in compute shader computations into a normal gdshader #6989
- Dispatch a draw call with RenderingDevice buffers serving as the mesh data like the Unity's [Graphics.DrawProceduralIndirect](Graphics.DrawProceduralIndirect)

Then they would be game-changing useful and allow things that truly harness the power of GPU's, like implementing compute shader based tesselation and virtual geometry.

**Comment by gamma-delta** [comment]
Hi Erick, I am LITERALLY doing exactly the same thing right now, and I've been chasing the problem for several days without any luck. Like you, I can get vertices onto the GPU in a compute shader, but I can't figure out how to get it to actually draw it without going through the CPU. If you figure it out please tell me!

**Comment by erickweil** [comment]
> Hi Erick, I am LITERALLY doing exactly the same thing right now, and I've been chasing the problem for several days without any luck. Like you, I can get vertices onto the GPU in a compute shader, but I can't figure out how to get it to actually draw it without going through the CPU. If you figure it out please tell me!

Well, the 'solution' I ended up isn't ideal, I figured out how to bind textures generated by the compute shader into a normar custom shader material, then my custom shader does texel fetches on the texture to get the vertex data. The only problem is the memory overhead because for it to work I needed a dummy vertex array and indices so that the vertex counts worked correctly, otherwise the shader wouldn't know how many vertices and what triangles to draw, since there is no way to do a drawProceduralIndirect like Unity does, obtaining indices info from buffers. 

So my half baked solution leaves a lot of memory unused since I need to allocate the maximum size I could get in the generating phase and in the shader drop the vertices that overflow the actually generated count. Also idk but I think reading from textures aren't as efficient as from structured buffers (if they were supported)

You can see it in my github: https://github.com/erickweil/GodotTests

**Comment by gamma-delta** [comment]
Cool!

I was recommended by a friend that it might be possible to use an atomic counter to "append" stuff to the buffer. IIRC, each tetrahedron can get sliced into... 0, 16, 24, or 32 triangles? I don't remember. Point being, it might be able to save some space by using an atomic counter and banking on how every tetrahedron probably isn't going to require the maximum number of output triangles. If one tetrahedron outputs 0 triangles, that leaves space for another tetrahedron to output many triangles.

**Comment by fire** [comment]
See also https://github.com/AIFanatic/three-nanite.

> An attempt at reproducing a dynamic LOD in threejs similarly to unreal's nanite. Very far from it but nonetheless a start. For now it clusters a mesh (meshlets), then groups adjacent clusters into a group, merges the mesh (shared vertices), performs mesh simplification to half the triangles in the mesh (max 128) and finally it splits it into 2 (should be N/2).

**Comment by fire** [comment]
@erickweil Maybe you write a proposal on adding to the Godot Engine api to obtain indices info from buffers for the dummy vertex array and indices problem?

**Comment by Radivarig** [comment]
Tessellation is being added to Terrain3D here https://github.com/TokisanGames/Terrain3D/pull/747

It would be useful to have this option because it can simplify mesh and textures authoring.

**Comment by Calinou** [comment]
> Tessellation is being added to Terrain3D here https://github.com/TokisanGames/Terrain3D/pull/747

Note that this is not actual hardware tesselation (using the tesselation control and evaluation stages), but vertex displacement in the shader. No new vertices are being created in the shader here.

The RenderingServer/RenderingDevice APIs do expose these hardware shader stages, but they aren't available in the Godot shader language.
